README.md
[![PyPI version fury.io](https://badge.fury.io/py/agrometeo-geopy.svg)](https://pypi.python.org/pypi/agrometeo-geopy/)
[![Documentation Status](https://readthedocs.org/projects/agrometeo-geopy/badge/?version=latest)](https://agrometeo-geopy.readthedocs.io/en/latest/?badge=latest)
[![CI/CD](https://github.com/martibosch/agrometeo-geopy/actions/workflows/dev.yml/badge.svg)](https://github.com/martibosch/agrometeo-geopy/blob/main/.github/workflows/dev.yml)
[![pre-commit.ci status](https://results.pre-commit.ci/badge/github/martibosch/agrometeo-geopy/main.svg)](https://results.pre-commit.ci/latest/github/martibosch/agrometeo-geopy/main)
[![codecov](https://codecov.io/gh/martibosch/agrometeo-geopy/branch/main/graph/badge.svg?token=hKoSSRn58a)](https://codecov.io/gh/martibosch/agrometeo-geopy)
[![GitHub license](https://img.shields.io/github/license/martibosch/agrometeo-geopy.svg)](https://github.com/martibosch/agrometeo-geopy/blob/main/LICENSE)
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/martibosch/agrometeo-geopy/HEAD?labpath=docs%2Fusage.ipynb)

# agrometeo-geopy

Pythonic interface to access [agrometeo.ch](https://agrometeo.ch) data.

## Installation

The agrometeo-geopy library requires [geopandas](https://geopandas.org/), which can be installed using conda/mamba as in:

```shell
conda install -c conda-forge geopandas
```

Then, agrometeo-geopy can be installed using pip:

```shell
pip install agrometeo-geopy
```

## Overview

```python
import agrometeo as agm

start_date = "2021-08-13"
end_date = "2021-08-16"

agm_ds = agm.AgrometeoDataset(region="Canton de Genève")
ts_df = agm_ds.get_ts_df(start_date=start_date, end_date=end_date)
ts_df
```

<div>
    <div class="wy-table-responsive"><table border="1" class="dataframe docutils">
            <thead>
                <tr style="text-align: right;">
                    <th>name</th>
                    <th>DARDAGNY</th>
                    <th>LA-PLAINE</th>
                    <th>SATIGNY</th>
                    <th>PEISSY</th>
                    <th>ANIERES</th>
                    <th>LULLY</th>
                    <th>LULLIER</th>
                    <th>BERNEX</th>
                    <th>TROINEX</th>
                    <th>MEINIER</th>
                </tr>
                <tr>
                    <th>time</th>
                    <th></th>
                    <th></th>
                    <th></th>
                    <th></th>
                    <th></th>
                    <th></th>
                    <th></th>
                    <th></th>
                    <th></th>
                    <th></th>
                </tr>
            </thead>
            <tbody>
                <tr>
                    <th>2021-08-13 00:00:00</th>
                    <td>19.3</td>
                    <td>17.8</td>
                    <td>18.5</td>
                    <td>17.9</td>
                    <td>20.6</td>
                    <td>18.4</td>
                    <td>20.3</td>
                    <td>18.6</td>
                    <td>19.4</td>
                    <td>25.8</td>
                </tr>
                <tr>
                    <th>2021-08-13 00:10:00</th>
                    <td>19.6</td>
                    <td>17.9</td>
                    <td>18.4</td>
                    <td>17.7</td>
                    <td>20.0</td>
                    <td>18.3</td>
                    <td>19.6</td>
                    <td>18.7</td>
                    <td>19.1</td>
                    <td>28.6</td>
                </tr>
                <tr>
                    <th>2021-08-13 00:20:00</th>
                    <td>19.0</td>
                    <td>17.7</td>
                    <td>18.2</td>
                    <td>17.6</td>
                    <td>19.4</td>
                    <td>18.4</td>
                    <td>19.1</td>
                    <td>18.7</td>
                    <td>19.2</td>
                    <td>24.1</td>
                </tr>
                <tr>
                    <th>2021-08-13 00:30:00</th>
                    <td>18.3</td>
                    <td>18.0</td>
                    <td>18.1</td>
                    <td>17.4</td>
                    <td>19.1</td>
                    <td>18.3</td>
                    <td>19.1</td>
                    <td>18.6</td>
                    <td>18.9</td>
                    <td>22.5</td>
                </tr>
                <tr>
                    <th>2021-08-13 00:40:00</th>
                    <td>18.7</td>
                    <td>18.0</td>
                    <td>18.1</td>
                    <td>17.6</td>
                    <td>19.1</td>
                    <td>18.0</td>
                    <td>19.0</td>
                    <td>18.7</td>
                    <td>18.5</td>
                    <td>21.5</td>
                </tr>
                <tr>
                    <th>...</th>
                    <td>...</td>
                    <td>...</td>
                    <td>...</td>
                    <td>...</td>
                    <td>...</td>
                    <td>...</td>
                    <td>...</td>
                    <td>...</td>
                    <td>...</td>
                    <td>...</td>
                </tr>
                <tr>
                    <th>2021-08-16 23:10:00</th>
                    <td>17.5</td>
                    <td>17.8</td>
                    <td>17.3</td>
                    <td>16.9</td>
                    <td>17.9</td>
                    <td>17.6</td>
                    <td>17.3</td>
                    <td>17.2</td>
                    <td>17.9</td>
                    <td>22.2</td>
                </tr>
                <tr>
                    <th>2021-08-16 23:20:00</th>
                    <td>17.4</td>
                    <td>17.9</td>
                    <td>17.4</td>
                    <td>17.1</td>
                    <td>17.9</td>
                    <td>17.6</td>
                    <td>17.3</td>
                    <td>17.2</td>
                    <td>18.0</td>
                    <td>22.0</td>
                </tr>
                <tr>
                    <th>2021-08-16 23:30:00</th>
                    <td>17.2</td>
                    <td>17.9</td>
                    <td>17.5</td>
                    <td>17.3</td>
                    <td>17.8</td>
                    <td>17.6</td>
                    <td>17.3</td>
                    <td>17.3</td>
                    <td>18.0</td>
                    <td>21.7</td>
                </tr>
                <tr>
                    <th>2021-08-16 23:40:00</th>
                    <td>17.2</td>
                    <td>17.9</td>
                    <td>17.7</td>
                    <td>17.1</td>
                    <td>17.7</td>
                    <td>17.4</td>
                    <td>17.2</td>
                    <td>17.1</td>
                    <td>18.1</td>
                    <td>21.9</td>
                </tr>
                <tr>
                    <th>2021-08-16 23:50:00</th>
                    <td>17.1</td>
                    <td>17.8</td>
                    <td>17.5</td>
                    <td>17.1</td>
                    <td>17.7</td>
                    <td>17.3</td>
                    <td>17.1</td>
                    <td>17.1</td>
                    <td>18.1</td>
                    <td>22.0</td>
                </tr>
            </tbody>
    </table></div>
    <p>576 rows × 10 columns</p>
</div>

See [the user guide](https://agrometeo-geopy.readthedocs.io/en/latest/usage.html) for more details.

## Acknowledgements

- This package was created with [Cookiecutter](https://github.com/audreyr/cookiecutter) and the [zillionare/cookiecutter-pypackage](https://github.com/zillionare/cookiecutter-pypackage) project template.


HISTORY.md
# History

## 0.2.0 (2023-04-10)

### Feat

- bump version to 0.2.0
- support multiple vars; station_id_col as setting
- sort_index in ts_df

### Fix

- avoid AttributeError when no contextily and add_basemap is True

### Refactor

- rename station -> stations in constants for consistency

## 0.1.3 (2023-04-03)

- fix: set columns based on str replace

## 0.1.2 (2022-03-21)

- fix: missing default plot settings; added plotting tests
- fix: geometry column in `ts_gdf`

## 0.1.1 (2022-03-21)

- fix: drop region arg in get_ts_gdf, forward args to get_ts_df
- fixed "all" imports in core and plotting

## 0.1.0 (2022-03-10)

- First release on PyPI.


tests/__init__.py
"""Unit test package for agrometeo_geopy."""


tests/test_agrometeo.py
#!/usr/bin/env python
"""Tests for `agrometeo` package."""
# pylint: disable=redefined-outer-name
import numpy as np

import agrometeo as agm


def test_agrometeo():
    # test core functions
    agm_ds = agm.AgrometeoDataset(region="Pully, Switzerland")
    assert agm_ds.CRS is not None
    num_stations = len(agm_ds.stations_gdf)
    assert num_stations >= 1
    start_date = "2022-03-22"
    end_date = "2022-03-23"
    # test that variable can be an ECV following the meteostations-geopy nomenclature, a
    # variable name following the agrometeo nomenclature and a variable code (as str or
    # int) following the agrometeo nomenclature
    for variable in ["temperature", "Precipitation", "1", 1]:
        ts_df = agm_ds.get_ts_df(
            variable=variable, start_date=start_date, end_date=end_date
        )
        assert len(ts_df.columns) == num_stations
        ts_gdf = agm_ds.get_ts_gdf(
            variable=variable, start_date=start_date, end_date=end_date
        )
        assert len(ts_gdf) == num_stations
        assert ts_gdf["geometry"].isna().sum() == 0

    # test plotting
    # use `add_basemap=False` to avoid having to mock contextily's requests
    ax = agm.plot_temperature_map(ts_gdf, add_basemap=False)
    assert len(ax.get_title()) > 0
    ax = agm.plot_temperature_map(ts_gdf, title=False, add_basemap=False)
    assert len(ax.get_title()) == 0
    ax = agm.plot_temperature_map(ts_gdf, title="some-title", add_basemap=False)
    assert len(ax.get_title()) > 0
    assert len(ax.collections[0].get_array()) == num_stations
    ts_columns = ts_gdf.columns.drop("geometry")
    axes = [
        agm.plot_temperature_map(
            ts_gdf,
            dt=dt,
            add_basemap=False,
        )
        for dt in [ts_columns[0], ts_columns[-1]]
    ]
    assert not np.array_equal(
        axes[0].collections[0].get_array(), axes[1].collections[0].get_array()
    )
    # test other args
    agm.plot_temperature_map(ts_gdf, add_basemap=False, plot_kws={"cmap": "Spectral"})
    agm.plot_temperature_map(ts_gdf, add_basemap=False, append_axes_kws={"pad": 0.4})


tests/test_agrometeo_geopy.py
#!/usr/bin/env python
"""Tests for `agrometeo_geopy` package."""
# pylint: disable=redefined-outer-name

import pytest


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string
    del response


agrometeo/base.py
"""Base abstract classes for meteo station datasets."""
import abc
import logging

import geopandas as gpd
import pandas as pd
from fiona.errors import DriverError

try:
    import osmnx as ox
except ImportError:
    ox = None

from . import settings

__all__ = ["MeteoStationDataset"]


def _process_region_arg(region=None, geocode_to_gdf_kws=None):
    if region is not None:
        if isinstance(region, gpd.GeoSeries):
            # if we have a GeoSeries, convert it to a GeoDataFrame so that we can use
            # the same code
            region = gpd.GeoDataFrame(geometry=region)
        elif not isinstance(region, gpd.GeoDataFrame):
            # at this point, we assume that this is either file-like or a Nominatim
            # query
            try:
                region = gpd.read_file(region)
            except DriverError:
                if ox is None:
                    logging.warning(
                        """
Using a Nominatim query as `region` argument requires the osmnx package. You can install
it using conda or pip. See https://github.com/geopandas/geopandas.
"""
                    )
                    return

                if geocode_to_gdf_kws is None:
                    geocode_to_gdf_kws = {}
                region = ox.geocode_to_gdf(region, **geocode_to_gdf_kws).iloc[:1]

    return region


def _long_ts_df(ts_df, station_id_name, time_name, value_name):
    """Transform time series data frame from wide (default) to long format."""
    return pd.melt(
        ts_df.reset_index(),
        id_vars=time_name,
        var_name=station_id_name,
        value_name=value_name,
    )


class MeteoStationDataset(metaclass=abc.ABCMeta):
    """Meteo station dataset."""

    def __init__(
        self,
        *,
        region=None,
        crs=None,
        stations_id_name=None,
        time_name=None,
        geocode_to_gdf_kws=None,
    ):
        """
        Initialize an meteo station dataset.

        Parameters
        ----------
        region : str, list-like, geopandas.GeoSeries, geopandas.GeoDataFrame, geometric
                 object, file-like object or pathlib.Path object, optional
            This can either be:

            * A string with a Nominatim query to geocode
            * A list-like with the west, south, east and north bounds
            * A geopandas geo-series or geo-data frame
            * A geometric object, e.g., shapely geometry
            * A filename or URL, a file-like object opened in binary ('rb') mode, or a
              Path object that will be passed to `geopandas.read_file`.
        """
        self.region = _process_region_arg(
            region=region, geocode_to_gdf_kws=geocode_to_gdf_kws
        ).to_crs(self.CRS)
        if stations_id_name is None:
            stations_id_name = settings.STATIONS_ID_NAME
        self.stations_id_name = stations_id_name
        if time_name is None:
            time_name = settings.TIME_NAME
        self.time_name = time_name

    @property
    @abc.abstractmethod
    def CRS(self):  # pylint: disable=invalid-name
        """CRS of the data source."""
        pass

    @abc.abstractproperty
    def stations_gdf(self):
        """Geo-data frame with stations data."""
        pass

    @abc.abstractmethod
    def get_ts_df(self, *args, **kwargs):
        """
        Get time series data frame.

        Returns
        -------
        ts_df : pd.DataFrame
            Data frame with a time series of meaurements (rows) at each station
            (columns).
        """
        pass

    @abc.abstractmethod
    def get_ts_gdf(self, *args, **kwargs):
        """
        Get time series geo-data frame.

        Returns
        -------
        ts_gdf : gpd.GeoDataFrame
            Geo-data frame with a time series of meaurements (columns) at each station
            (rows), with an additional geometry column with the stations' locations.
        """
        pass


agrometeo/core.py
"""Agrometeo."""
import datetime
import warnings
from os import path

import geopandas as gpd
import numpy as np
import pandas as pd
import requests
from shapely.errors import ShapelyDeprecationWarning
from shapely.geometry import Point

from . import base, settings

__all__ = ["AgrometeoDataset"]

# API endpoints
BASE_URL = "https://www.agrometeo.ch/backend/api"
# the agrometeo API uses integer ids to identify stations, which can be retrieved from
# the "id" column of the stations dataframe
STATIONS_API_ID_COL = "id"
STATIONS_API_ENDPOINT = path.join(BASE_URL, "stations")
VARIABLES_NAME_COL = "name.en"
VARIABLES_CODE_COL = "id"
VARIABLES_API_ENDPOINT = path.join(BASE_URL, "sensors")
# note that there is a `stations/near` endpoint as in:
# stations/near?latitude=46.433&longitude=6.9114
METEO_DATA_API_ENDPOINT = path.join(BASE_URL, "meteo/data")

# useful constants
LONLAT_CRS = "epsg:4326"
LV03_CRS = "epsg:21781"
GEOM_COL_DICT = {LONLAT_CRS: ["long_dec", "lat_dec"], LV03_CRS: ["lon_ch", "lat_ch"]}
API_DT_FMT = "%Y-%m-%d"
SJOIN_PREDICATE = "within"
SCALE = "none"
MEASUREMENT = "avg"


class AgrometeoDataset(base.MeteoStationDataset):
    """Agrometeo dataset."""

    def __init__(
        self,
        *,
        region=None,
        stations_id_name=None,
        time_name=None,
        crs=None,
        geocode_to_gdf_kws=None,
        sjoin_kws=None,
    ):
        """
        Initialize an Agrometeo dataset.

        Parameters
        ----------
        region : str, list-like, geopandas.GeoSeries, geopandas.GeoDataFrame, geometric
                 object, file-like object or pathlib.Path object, optional
            This can either be:

            * A string with a Nominatim query to geocode
            * A list-like with the west, south, east and north bounds
            * A geopandas geo-series or geo-data frame
            * A geometric object, e.g., shapely geometry
            * A filename or URL, a file-like object opened in binary ('rb') mode, or a
              Path object that will be passed to `geopandas.read_file`.
        """
        # ACHTUNG: need to define the CRS before calling the parent's init
        if crs is None:
            crs = LONLAT_CRS
        self.crs = crs

        super().__init__(
            region=region,
            stations_id_name=stations_id_name,
            time_name=time_name,
            geocode_to_gdf_kws=geocode_to_gdf_kws,
        )

        if sjoin_kws is None:
            sjoin_kws = {}
        self.sjoin_kws = sjoin_kws

    @property
    def CRS(self):  # pylint: disable=invalid-name
        """CRS of the data source."""
        return self.crs

    @property
    def stations_gdf(self):
        """Station geo-data frame."""
        try:
            return self._stations_gdf
        except AttributeError:
            geom_cols = GEOM_COL_DICT[self.crs]
            response = requests.get(STATIONS_API_ENDPOINT)
            stations_df = pd.json_normalize(response.json()["data"])
            # it is fine to filter out this ShapelyDeprecationWarning, see
            # https://shapely.readthedocs.io/en/latest/migration.html
            # #creating-numpy-arrays-of-geometry-objects
            with warnings.catch_warnings():
                warnings.filterwarnings("ignore", category=ShapelyDeprecationWarning)
                stations_gdf = gpd.GeoDataFrame(
                    stations_df.drop(geom_cols, axis=1),
                    geometry=stations_df[geom_cols]
                    .astype(np.float64)
                    .apply(lambda xy_ser: Point(xy_ser[0], xy_ser[1]), axis=1),
                    crs=self.crs,
                )

            _sjoin_kws = self.sjoin_kws.copy()
            predicate = _sjoin_kws.pop("predicate", SJOIN_PREDICATE)
            stations_gdf = stations_gdf.sjoin(
                self.region.to_crs(stations_gdf.crs), predicate=predicate, **_sjoin_kws
            )
            # stations_gdf.index.name = self.stations_id_name

            self._stations_gdf = stations_gdf

            return self._stations_gdf

    @property
    def variables_df(self):
        """Variables data frame."""
        try:
            return self._variables_df
        except AttributeError:
            response = requests.get(VARIABLES_API_ENDPOINT)
            variables_df = pd.json_normalize(response.json()["data"])

            # ACHTUNG: need to strip strings, at least in variables name column. Note
            # that *it seems* that the integer type of variable code column is inferred
            # correctly
            variables_df[VARIABLES_NAME_COL] = variables_df[
                VARIABLES_NAME_COL
            ].str.strip()

            self._variables_df = variables_df

            return self._variables_df

    def _get_region_data(self, variable_code, start_date, end_date, scale, measurement):
        # use a variable for the station ids just to keep the line below shorter
        _stations_ids = self.stations_gdf[STATIONS_API_ID_COL].astype(str)
        request_url = f"{METEO_DATA_API_ENDPOINT}?" + "&".join(
            [
                f"from={start_date}",
                f"to={end_date}",
                f"scale={scale}",
                f"sensors={variable_code}%3A{measurement}",
                f"stations={'%2C'.join(_stations_ids)}",
            ]
        )

        return requests.get(request_url)

    def get_ts_df(
        self,
        variable,
        start_date,
        end_date,
        *,
        scale=None,
        measurement=None,
        stations_id_col=None,
    ):
        """
        Get time series data frame.

        Parameters
        ----------
        variable : str or int
            Target variable, which can be either an agrometeo variable code (integer or
            string), an essential climate variable (ECV) following the
            meteostations-geopy nomenclature (string), or an agrometeo variable name
            (string).
        start_date, end_date : str or datetime
            String in the "YYYY-MM-DD" format or datetime instance, respectively
            representing the start and end of the  requested data period.
        scale : None or {"hour", "day", "month", "year"}, default None
            Temporal scale of the measurements. The default value of None returns the
            finest scale, i.e., 10 minutes.
        measurement : {"min", "avg", "max"}, default "avg"
            Whether the measurement values correspond to the minimum, average or maximum
            value for the required temporal scale. Ignored if `scale` is None.
        stations_id_col : str, optional
            Column of `stations_gdf` that will be used in the returned data frame to
            identify the stations. If None, the value from
            `settings.DEFAULT_STATIONS_ID_COL` will be used.

        Returns
        -------
        ts_df : pd.DataFrame
            Data frame with a time series of meaurements (rows) at each station
            (columns).
        """
        # process the variable arg
        # variable is a string that can be either:
        # a) an agrometeo variable code
        # b) an essential climate variable (ECV) following the meteostations-geopy
        # nomenclature
        # c) an agrometeo variable name
        if isinstance(variable, int) or variable.isdigit():
            # case a: if variable is an integer, assert that it is a valid variable code
            variable_code = int(variable)
            if variable_code not in self.variables_df[VARIABLES_CODE_COL].values:
                raise ValueError(
                    f"variable {variable} is not a valid agrometeo variable code"
                )
        else:
            # case b or c: if variable is an ECV, it will be a key in the ECV_DICT so
            # the agrometeo variable code can be retrieved directly, otherwise we
            # assume that variable is an agrometeo variable name
            agm_variable_name = settings.ECV_DICT.get(variable, variable)
            variable_code = self.variables_df.loc[
                self.variables_df[VARIABLES_NAME_COL] == agm_variable_name,
                VARIABLES_CODE_COL,
            ].item()
        # process date args
        if isinstance(start_date, datetime.datetime):
            start_date = start_date.strftime(API_DT_FMT)
        if isinstance(end_date, datetime.datetime):
            end_date = end_date.strftime(API_DT_FMT)
        # process scale and measurement args
        if scale is None:
            # the API needs it to be lowercase
            scale = SCALE
        if measurement is None:
            measurement = MEASUREMENT
        # process the stations_id_col arg
        if stations_id_col is None:
            stations_id_col = settings.DEFAULT_STATIONS_ID_COL

        # query the API
        response = self._get_region_data(
            variable_code, start_date, end_date, scale, measurement
        )

        # parse the response as a data frame
        ts_df = pd.json_normalize(response.json()["data"]).set_index("date")
        ts_df.index = pd.to_datetime(ts_df.index)
        ts_df.index.name = self.time_name
        # ts_df.columns = self.stations_gdf[STATIONS_ID_COL]
        # ACHTUNG: note that agrometeo returns the data indexed by keys of the form
        # "{station_id}_{variable_code}_{measurement}", so to properly set the columns
        # as the desired station identifier (e.g., "id" or "name") we need to first get
        # the ids and then get (loc) the station data from the stations_gdf.
        ts_df.columns = self.stations_gdf.set_index(STATIONS_API_ID_COL).loc[
            ts_df.columns.str.replace(f"_{variable_code}_{measurement}", "").astype(
                self.stations_gdf[STATIONS_API_ID_COL].dtype
            )
        ][stations_id_col]
        ts_df = ts_df.apply(pd.to_numeric, axis=1)

        return ts_df.sort_index()

    def get_ts_gdf(
        self,
        variable,
        start_date,
        end_date,
        *,
        scale=None,
        measurement=None,
        stations_id_col=None,
    ):
        """
        Get time series geo-data frame.

        Parameters
        ----------
        variable : str or int
            Target variable, which can be either an agrometeo variable code (integer or
            string), an essential climate variable (ECV) following the
            meteostations-geopy nomenclature (string), or an agrometeo variable name
            (string).
        start_date, end_date : str or datetime
            String in the "YYYY-MM-DD" format or datetime instance, respectively
            representing the start and end of the  requested data period.
        scale : None or {"hour", "day", "month", "year"}, default None
            Temporal scale of the measurements. The default value of None returns the
            finest scale, i.e., 10 minutes.
        measurement : {"min", "avg", "max"}, default "avg"
            Whether the measurement values correspond to the minimum, average or maximum
            value for the required temporal scale. Ignored if `scale` is None.
        stations_id_col : str, optional
            Column of `stations_gdf` that will be used in the returned data frame to
            identify the stations. If None, the value from
            `settings.DEFAULT_STATIONS_ID_COL` is used.

        Returns
        -------
        ts_gdf : gpd.GeoDataFrame
            Geo-data frame with a time series of meaurements (columns) at each station
            (rows), with an additional geometry column with the stations' locations.
        """
        ts_gdf = gpd.GeoDataFrame(
            self.get_ts_df(
                variable,
                start_date,
                end_date,
                scale=scale,
                measurement=measurement,
                stations_id_col=stations_id_col,
            ).T
        )
        # get the geometry from stations_gdf
        ts_gdf["geometry"] = self.stations_gdf.set_index(ts_gdf.index.name).loc[
            ts_gdf.index
        ]["geometry"]
        # sort the timestamp columns
        ts_columns = ts_gdf.columns.drop("geometry")
        ts_gdf = ts_gdf[sorted(ts_columns) + ["geometry"]]

        return ts_gdf


agrometeo/settings.py
"""Settings."""
# core
STATIONS_ID_NAME = "station_id"
TIME_NAME = "time"

# plotting
PLOT_CMAP = "coolwarm"
PLOT_LEGEND = True
PLOT_ATTRIBUTION = False
PLOT_LEGEND_POSITION = "right"
PLOT_LEGEND_SIZE = "2.5%"
PLOT_LEGEND_PAD = 0.2
PLOT_TITLE = True
PLOT_ADD_BASEMAP = True

# agrometeo specific
DEFAULT_STATIONS_ID_COL = "name"
# https://public.wmo.int/en/programmes/global-climate-observing-system/essential-climate-variables
ECVS = [
    "precipitation",
    "pressure",
    "surface_radiation_longwave",
    "surface_radiation_shortwave",
    "surface_wind_speed",
    "surface_wind_direction",
    "temperature",
    "water_vapour",
]
# agrometeo sensors
# 42                       Leaf moisture III
# 43     Voltage of internal lithium battery
# 1              Temperature 2m above ground
# 4                       Relative humidity
# 6                           Precipitation
# 15              Intensity of precipitation
# 7                            Leaf moisture
# 11                         Solar radiation
# 41                           Solar Energie
# 9                          Avg. wind speed
# 14                         Max. wind speed
# 8                           Wind direction
# 22                       Temperature +10cm
# 12                    Luxmeter after Lufft
# 10                                ETP-Turc
# 24                              ETo-PenMon
# 13                               Dew point
# 18                       Real air pressure
# 2                    Soil temperature +5cm
# 19                  Soil temperature -20cm
# 3                   Soil temperature -10cm
# 5                       Soil moisture -5cm
# 20                   Pressure on sea level
# 17                        Leaf moisture II
# 25                     Soil moisture -30cm
# 26                     Soil moisture -50cm
# 39                                  unused
# 33                 Temperature in leafzone
# 32                         battery voltage
# 21                         min. wind speed
# 23                        Temperatur +20cm
# 27                  Temperatur in Pflanze1
# 28                  Temperatur in Pflanze1
# 29                                    UVAB
# 30                                     UVA
# 31                                     UAB
# 34                Air humidity in leafzone
# 35             Photosyth. active radiation
# 36                  Soil temperature -10cm
# 37                Temperatur 2m unbelüftet
# 38           elative Luftfeuchtigkeit +5cm
# 40                     Precip. Radolan Day
# 100                                   Hour
# 101                                   Year
# 102                            Day of year
# 103                           Degree hours
# 104                 Density of sporulation
# 105                           Leaf surface

ECV_DICT = {
    "precipitation": "Precipitation",
    "pressure": "Real air pressure",
    "surface_radiation_shortwave": "Solar radiation",
    "surface_wind_speed": "Avg. wind speed",
    "surface_wind_direction": "Wind direction",
    "temperature": "Temperature 2m above ground",
    "water_vapour": "Relative humidity",
}
